ql/models/shortrate/calibrationhelpers/swaptionhelper.cpp

```diff
@@ -168,32 +183,37 @@ namespace QuantLib {
                              new DiscountingSwapEngine(termStructure_, false));
 
         Swap::Type type = Swap::Receiver;
-
-        VanillaSwap temp(Swap::Receiver, nominal_,
-                            fixedSchedule, 0.0, fixedLegDayCounter_,
-                            floatSchedule, index_, 0.0, floatingLegDayCounter_);
-        temp.setPricingEngine(swapEngine);
-        Real forward = temp.fairRate();
-        if(strike_ == Null<Real>()) {
-            exerciseRate_ = forward;
+        ext::shared_ptr<Exercise> exercise(new EuropeanExercise(exerciseDate));
+        FixedVsFloatingSwap* temp;
+        if (onIndex) {
+            // construct ois swap
+            temp = new OvernightIndexedSwap(Swap::Receiver, nominal_, fixedSchedule, 0.0,
+                                            fixedLegDayCounter_, onIndex, 0.0, 0, Following, Calendar(),
+                                            true, averagingMethod_);
+        } else {
+            temp = new VanillaSwap(Swap::Receiver, nominal_, fixedSchedule, 0.0, fixedLegDayCounter_,
+                                   floatSchedule, index_, 0.0, floatingLegDayCounter_);
         }
-        else {
+        temp->setPricingEngine(swapEngine);
+        Real forward = temp->fairRate();
+        delete temp;
```

Comment:
This is going to leak if `fairRate` happens to throw an exception.  I'd make `temp` a `shared_ptr` or `unique`ptr`.

Reply:
fixed.